distiller/extraction/analyze.py

```diff
@@ -60,3 +61,73 @@ def analyze_doc(page_doc, audit, page_number):
             page_number=page_number,
         )
     return None
+
+
+def output_as_csv(results):
+    """
+    Output analyzed results in a format suitable for CSV. Audit
+    number, finding, CAP, and page number columns are always present.
+    Additional finding keywords, if any, are added as extra columns.
+    """
+    from functools import reduce
+
+    base_header = [
+        "audit number",
+        "finding data",
+        "corrective action plan",
+        "page number",
+    ]
+    headers = reduce(
+        lambda accum, x: accum | set(x["finding_data"].keys()), results, set()
+    )
+    headers.remove("Finding")  # already present in base header
+    headers.remove("Page number")  # already present in base header
+    yield base_header + list(headers)
+
+    for result in results:
+        audit = result["audit"]
+        page_number = result["page_number"]
+        finding = result["finding_data"].get("Finding", "")
+        cap = result["cap_data"].get("Plan", "")
+        rest = map(lambda h: result["finding_data"].get(h, ""), headers)
+        yield [audit, finding, cap, page_number] + list(rest)
+
+
+if __name__ == "__main__":
+    import argparse
+    import csv
+    import pickle
+
+    from distiller.gateways import files
+
+    parser = argparse.ArgumentParser(
+        description="Extract audit data from the given PDF."
+    )
+    parser.add_argument("filename")
+    parser.add_argument("--csv", help="store output to a CSV file")
+    parser.add_argument("--pickle", help="store output to a pickle file")
+    args = parser.parse_args()
+
+    pdf = files.input_file(args.filename, mode="rb")
+    errors = pdf_utils.errors(pdf)
+    if errors:
+        print(f"Could not read file {args.filename}: {errors}. Bailing out.")
+        sys.exit(1)
+
+    print(f"processing {args.filename}")
+    processor = nlp.setup()
+    audit_results = analyze(processor, pdf, silent=True)
+
+    if args.pickle:
+        with open(args.pickle, mode="wb") as fd:
+            pickle.dump(audit_results, fd)
+    elif args.csv:
+        with open(args.csv, mode="w") as fd:
```

Comment:
You could also use `files.output_file` if you want to support writing to S3.

Reply:
Great idea, thanks. Updated.